### ISSUE TITLE
Lots of "Use of uninitialized value" warnings after biber update

### ISSUE THREAD

**Issue by RalfJung** [plk/biber#403]
I updated biber to 2.17, and now I am getting many warnings (thousands of them for large bib files, which significantly slows down biber itself) from files that were clean before. The warnings all look like this:
```
Use of uninitialized value $opt in hash element at /usr/share/perl5/Biber/Config.pm line 977.
Use of uninitialized value within %Biber::Config::CONFIG_OPTSCOPE_BIBLATEX in hash dereference at /usr/share/perl5/Biber/Config.pm line 977.
```
Sometimes the line numbers are different (I have seen 967, 972, 977).

*Any* use of biber causes that warning for me, so here is an arbitrary example:
```
\documentclass[10pt,a4paper]{article}
\usepackage[backend=biber]{biblatex}

\addbibresource{bib.bib}

\begin{document}

\fullcite{iris}

\end{document}
```
With `bib.bib` being
```
@inproceedings{iris,
  author    = {Ralf Jung and
               David Swasey and
               Filip Sieczkowski and
               Kasper Svendsen and
               Aaron Turon and
               Lars Birkedal and
               Derek Dreyer},
  title     = {{Iris}: Monoids and Invariants as an Orthogonal Basis for Concurrent
               Reasoning},
  booktitle = {{POPL}},
  year      = {2015},
  doi       = {10.1145/2676726.2676980},
}
```

Biber (and the rest of texlive) are installed via Debian testing packages.

**Comment by moewew** [comment]
That's not something I can reproduce here (MikTeX on Windows 10 and TUG vanilla TeX live 2021 on Ubuntu 18.04).

Maybe it's the cache. Try deleting it as described in https://tex.stackexchange.com/q/140814/35864 and rerun.

Though I seem to recall that Debian package things differently, so your Biber may not even have a cache. In which case I suggest you get in touch with the Biber Debian maintainers, they might be in a better position to suss out what is going on here.

**Comment by RalfJung** [comment]
I get this problem even in a new empty directory with just those 2 files, so it's not the cache.

I'll make a Debian bugreport, thanks for the quick reaction!

**Comment by moewew** [comment]
The cache I was referring to is not a cache that caches your working directory or `.bib` files. It is a cache that contains the unpacked Biber binary (a little background is at https://tex.stackexchange.com/a/541494/35864). So in order to really find out if the cache is involved, you need to follow the steps in https://tex.stackexchange.com/q/140814/35864. But as I say, I think I once gathered that Debian does not use the pre-built binaries, so they might not have a cache at all.

**Comment by RalfJung** [comment]
(Yes, sorry for the initial misunderstanding about the cache. I have edited my above reply since initially posting it.)

**Comment by hpreusse** [comment]
> That's not something I can reproduce here (MikTeX on Windows 10 and TUG vanilla TeX live 2021 on Ubuntu 18.04).
> 
> Maybe it's the cache. Try deleting it as described in https://tex.stackexchange.com/q/140814/35864 and rerun.
> 
I'm able to reproduce the issue. Please note that Ubuntu 18.04 (and probably MikTeX on Win 10) has an older version of perl installed. Please be so kind to test it with the version of perl in Debian unstable / testing (5.34). Many thanks!

**Comment by moewew** [comment]
Since I tested the pre-built binaries, the Perl version available on my machines does not matter. The pre-built binaries bring their own compatible Perl.

I can't tell you which Perl version was used to build the x86_64 binaries for Linux (those are built by @plk), but I can tell you that I built the Linux x86_32 binaries with Perl 5.34 (fully updated at the time) and I did not encounter any such problems in all the tests I ran.

**Comment by hpreusse** [comment]
> I played around with a Debian Sid installation I installed today and found that
> 
>     * When I wanted to install Biber from GitHub/`master`, some dependencies were reported missing. I installed some of them, but did not manage to install all of them.
> 
Yes, this is correct. If you look at the [build log of the package](https://buildd.debian.org/status/fetch.php?pkg=biber&arch=all&ver=2.17-2&stamp=1644187528&raw=0), you'll notice that deps are always reported to be missing:

```
    !  Encode::EUCJPASCII is not installed
    !  Mozilla::CA is not installed
```
as both packages are not in Debian (yet). The first will be available at any time, the latter probably never, but I hope this does not cause issues.

>     * I then installed Biber from GitHub/`master` (ignoring the missing dependencies). That Biber worked without an issue.
>     * I removed the GitHub-installed Biber and the Debian one would still error (so the dependencies might not be the issue after all?).
> 
So, maybe we "build" biber the wrong way or something is missing: would you be so kind to have a look at our build log (link see above), maybe you have an idea, what we're doing wrong. The relevant part start at a line containing `dh_auto_configure -i`. Many thanks!

**Comment by dcpurton** [comment]
I also see this issue under Debian testing with `biber 2.17`.

The only difference I notice with the `biber` script produced from running `perl Build.PL` from the GitHub source and the system `/usr/bin/biber` is:

```diff
--- biber	2022-03-16 21:06:32.333171804 +1030
+++ /usr/bin/biber	2022-03-16 21:30:25.240899320 +1030
@@ -1,4 +1,4 @@
-#!/usr/bin/env perl
+#!/usr/bin/perl -w
 
 use v5.24;
 use strict;
```

If I change the interpreter of `/usr/bin/biber` to `#!/usr/bin/env perl` then everything works as expected.

**Comment by hpreusse** [comment]
So the main difference is that we call perl with option "-w". After removing the option the warnings go away. Not sure if we should solve the issue this way.

**Comment by plk** [comment]
Fixed in `biber` dev but those warnings can be ignored.

**Comment by moewew** [comment]
Down to two warnings now
```
Use of uninitialized value $tool in concatenation (.) or string at <PERLPATH>/Biber/Config.pm line 305.
Use of uninitialized value in string eq at <PERLPATH>/Biber/Internals.pm line 1507.
```

**Comment by hpreusse** [comment]
> Fixed in `biber` dev but those warnings can be ignored.

Thanks for the fix! Would be nice if the two remaining warnings mentioned by @moewew could be removed too.

**Comment by plk** [comment]
How are you getting these warnings? Running the test suite?

**Comment by moewew** [comment]
I run Biber on `test-dev` from https://sourceforge.net/projects/biblatex-biber/files/biblatex-biber/testfiles/ (changed the `.bcf` version to match the Biber dev version). I didn't run any more extensive tests, because I don't have TeX installed on the system I use to build Biber.

**Comment by plk** [comment]
Should be fixed now I believe. One was a fairly obscure edge case, the other was due to the fix put in for plk/biblatex-apa#152

**Comment by moewew** [comment]
The
```
Use of uninitialized value in string eq at <PERLPATH>/Biber/Internals.pm line 1507.
```
is still there for me.

All other warnings are gone.

**Comment by plk** [comment]
Should be fixed - needed to update the `test-dev.bcf`

**Comment by moewew** [comment]
Yeah, that's done it.

As I say, I don't have a proper TeX live on the machine I use to build Biber, so I can't really test things "in the wild". Maybe I'll manage to run some more tests at some point.

**Comment by MohamedaliS** [comment]
> Alright I can reproduce the warnings if I tell the perl interpreter to `-w`arn. Specifically I get
> 
> ```
> Use of uninitialized value in quotemeta at <PERLPATH>/Biber/Config.pm line 228.
> Use of uninitialized value $tool in concatenation (.) or string at <PERLPATH>/Biber/Config.pm line 307.
> Use of uninitialized value in string eq at <PERLPATH>/Biber/Internals.pm line 1507.
> Use of uninitialized value $opt in hash element at <PERLPATH>/Biber/Config.pm line 968.
> Use of uninitialized value within %Biber::Config::CONFIG_OPTSCOPE_BIBLATEX in hash dereference at <PERLPATH>/Biber/Config.pm line 968.
> Use of uninitialized value $opt in hash element at <PERLPATH>/Biber/Config.pm line 973.
> Use of uninitialized value within %Biber::Config::CONFIG_OPTSCOPE_BIBLATEX in hash dereference at <PERLPATH>/Biber/Config.pm line 973.
> Use of uninitialized value $opt in hash element at <PERLPATH>/Biber/Config.pm line 978.
> Use of uninitialized value within %Biber::Config::CONFIG_OPTSCOPE_BIBLATEX in hash dereference at <PERLPATH>/Biber/Config.pm line 978.
> Use of uninitialized value $opt in hash element at <PERLPATH>/Biber/Config.pm line 968.
> Use of uninitialized value within %Biber::Config::CONFIG_OPTSCOPE_BIBLATEX in hash dereference at <PERLPATH>/Biber/Config.pm line 968.
> Use of uninitialized value $opt in hash element at <PERLPATH>/Biber/Config.pm line 973.
> Use of uninitialized value within %Biber::Config::CONFIG_OPTSCOPE_BIBLATEX in hash dereference at <PERLPATH>/Biber/Config.pm line 973.
> Use of uninitialized value $opt in hash element at <PERLPATH>/Biber/Config.pm line 978.
> Use of uninitialized value within %Biber::Config::CONFIG_OPTSCOPE_BIBLATEX in hash dereference at <PERLPATH>/Biber/Config.pm line 978.
> Use of uninitialized value $opt in hash element at <PERLPATH>/Biber/Config.pm line 968.
> Use of uninitialized value within %Biber::Config::CONFIG_OPTSCOPE_BIBLATEX in hash dereference at <PERLPATH>/Biber/Config.pm line 968.
> Use of uninitialized value $opt in hash element at <PERLPATH>/Biber/Config.pm line 973.
> Use of uninitialized value within %Biber::Config::CONFIG_OPTSCOPE_BIBLATEX in hash dereference at <PERLPATH>/Biber/Config.pm line 973.
> Use of uninitialized value $opt in hash element at <PERLPATH>/Biber/Config.pm line 978.
> Use of uninitialized value within %Biber::Config::CONFIG_OPTSCOPE_BIBLATEX in hash dereference at <PERLPATH>/Biber/Config.pm line 978.
> ```
> 
> from the test `.bcf` from Sourceforge.
> 
> I had a look at the first warning and it looks like an undeclared option is being referenced. So this is something that should be looked at on the Biber side it appears to me.

I got the same error now in Ubuntu - Texstudio. How did you solve it?

**Comment by moewew** [comment]
Which version of Biber are you running? The warnings discussed here were resolved in version 2.18. If your Ubuntu is not the newest version and you installed TeX Live and Biber via `apt`, you're very likely not running the newest version of Biber - possibly one from before the fix. As discussed above though, the warnings are mostly harmless and can be ignored.

**Comment by YasserAderghal** [comment]
> I also see this issue under Debian testing with `biber 2.17`.
> 
> The only difference I notice with the `biber` script produced from running `perl Build.PL` from the GitHub source and the system `/usr/bin/biber` is:
> 
> ```diff
> --- biber	2022-03-16 21:06:32.333171804 +1030
> +++ /usr/bin/biber	2022-03-16 21:30:25.240899320 +1030
> @@ -1,4 +1,4 @@
> -#!/usr/bin/env perl
> +#!/usr/bin/perl -w
>  
>  use v5.24;
>  use strict;
> ```
> 
> If I change the interpreter of `/usr/bin/biber` to `#!/usr/bin/env perl` then everything works as expected.

using `#!/usr/bin/env perl` solved mine.
